zokrates_core/src/absy/from_ast.rs

```diff
@@ -0,0 +1,515 @@
+use absy;
+use imports;
+use types::Type;
+use zokrates_field::field::Field;
+use zokrates_pest_ast as pest;
+
+impl<'ast, T: Field> From<pest::File<'ast>> for absy::Prog<'ast, T> {
+    fn from(prog: pest::File<'ast>) -> absy::Prog<T> {
+        absy::Prog {
+            functions: prog
+                .functions
+                .into_iter()
+                .map(|f| absy::FunctionNode::from(f))
+                .collect(),
+            imports: prog
+                .imports
+                .into_iter()
+                .map(|i| absy::ImportNode::from(i))
+                .collect(),
+            imported_functions: vec![],
+        }
+    }
+}
+
+impl<'ast> From<pest::ImportDirective<'ast>> for absy::ImportNode {
+    fn from(import: pest::ImportDirective<'ast>) -> absy::ImportNode {
+        use absy::NodeValue;
+
+        imports::Import::new(import.source.value)
+            .alias(import.alias.map(|a| a.value))
+            .span(import.span)
+    }
+}
+
+impl<'ast, T: Field> From<pest::Function<'ast>> for absy::FunctionNode<'ast, T> {
+    fn from(function: pest::Function<'ast>) -> absy::FunctionNode<T> {
+        use absy::NodeValue;
+
+        let span = function.span;
+
+        let signature = absy::Signature::new()
+            .inputs(
+                function
+                    .parameters
+                    .clone()
+                    .into_iter()
+                    .map(|p| absy::ParameterNode::from(p).value.id.value.get_type())
+                    .collect(),
+            )
+            .outputs(
+                function
+                    .returns
+                    .clone()
+                    .into_iter()
+                    .map(|r| Type::from(r))
+                    .collect(),
+            );
+
+        absy::Function::<T> {
+            id: function.id.span.as_str(),
+            arguments: function
+                .parameters
+                .into_iter()
+                .map(|a| absy::ParameterNode::from(a))
+                .collect(),
+            statements: function
+                .statements
+                .into_iter()
+                .flat_map(|s| statements_from_statement(s))
+                .collect(),
+            signature,
+        }
+        .span(span)
+    }
+}
+
+impl<'ast> From<pest::Parameter<'ast>> for absy::ParameterNode<'ast> {
+    fn from(param: pest::Parameter<'ast>) -> absy::ParameterNode {
+        use absy::NodeValue;
+
+        let private = param
+            .visibility
+            .map(|v| match v {
+                pest::Visibility::Private(_) => true,
```

Comment:
I found it surprising that the visibilities are represented as structs in the zokrates_pest_ast. Since it's converted to bool here, it's fine as isolated from core.

Reply:
For reference, this is done because we don't want to parse this ourselves, so creating a rule and an associated struct works.